airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,269 @@
+from airbyte_cdk.sources.types import Record, StreamSlice, StreamState
+
+
+class Timer:
+    """
+    A simple timer class that measures elapsed time in seconds using a high-resolution performance counter.
+    """
+
+    def __init__(self) -> None:
+        self._start: Optional[int] = None
+
+    def start(self) -> None:
+        self._start = time.perf_counter_ns()
+
+    def finish(self) -> int:
+        if self._start:
+            return int((time.perf_counter_ns() - self._start) // 1e9)
+        else:
+            raise RuntimeError("Global substream cursor timer not started")
+
+
+class GlobalSubstreamCursor(DeclarativeCursor):
+    """
+    The GlobalSubstreamCursor is designed to track the state of substreams using a single global cursor.
+    This class is beneficial for streams with many partitions, as it allows the state to be managed globally
+    instead of per partition, simplifying state management and reducing the size of state messages.
+
+    This cursor is activated by setting the `global_substream_cursor` parameter for incremental sync.
+
+    Warnings:
+    - This class enforces a minimal lookback window for substream based on the duration of the previous sync to avoid losing records. This lookback ensures that any records added or updated during the sync are captured in subsequent syncs.
+    - The global cursor is updated only at the end of the sync. If the sync ends prematurely (e.g., due to an exception), the state will not be updated.
+    - When using the `incremental_dependency` option, the sync will progress through parent records, preventing the sync from getting infinitely stuck. However, it is crucial to understand the requirements for both the `global_substream_cursor` and `incremental_dependency` options to avoid data loss.
+    """
+
+    def __init__(self, stream_cursor: DeclarativeCursor, partition_router: PartitionRouter):
+        self._stream_cursor = stream_cursor
+        self._partition_router = partition_router
+        self._timer = Timer()
+        self._lock = threading.Lock()
+        self._slice_semaphore = threading.Semaphore(0)  # Start with 0, indicating no slices being tracked
+        self._all_slices_yielded = False
+        self._lookback_window: Optional[int] = None
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        """
+        Generate stream slices and ensure that the last slice is flagged appropriately.
+        This method uses a semaphore to track the processing of slices and ensures
+        that `close_slice` is called only after all slices have been processed.
+        """
+        previous_slice = None
+
+        slice_generator = (
+            StreamSlice(partition=partition, cursor_slice=cursor_slice)
+            for partition in self._partition_router.stream_slices()
+            for cursor_slice in self._stream_cursor.stream_slices()
+        )
+
+        for slice in slice_generator:
+            if previous_slice is not None:
```

Comment:
I'm not sure I get this condition. Why do we not yield the first slice?

Reply:
To set `_all_slices_yielded` to True before yielding the last slice